drivers/sensor/mb7040/mb7040.c

```diff
@@ -0,0 +1,176 @@
+	uint16_t distance_cm;
+	struct k_sem read_sem;
+#if MB7040_HAS_STATUS_GPIO
+	struct gpio_callback gpio_cb;
+#endif
+};
+
+struct mb7040_config {
+	struct i2c_dt_spec i2c;
+#if MB7040_HAS_STATUS_GPIO
+	struct gpio_dt_spec status_gpio;
+#endif
+};
+
+#if MB7040_HAS_STATUS_GPIO
+static void status_gpio_callback(const struct device *dev, struct gpio_callback *cb, uint32_t pins)
+{
+	struct mb7040_data *data = CONTAINER_OF(cb, struct mb7040_data, gpio_cb);
+	k_sem_give(&data->read_sem);
+}
+#endif
+
+static int mb7040_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	const struct mb7040_config *cfg = (struct mb7040_config *)dev->config;
+	struct mb7040_data *data = (struct mb7040_data *)dev->data;
+	uint8_t cmd = RANGE_CMD;
+	int ret;
+	uint8_t read_data[2];
+
+	if (chan != SENSOR_CHAN_DISTANCE && chan != SENSOR_CHAN_ALL) {
+		LOG_ERR("Sensor only supports distance");
+		return -EINVAL;
+	}
+
+	k_sem_reset(&data->read_sem);
+
+#if MB7040_HAS_STATUS_GPIO
+	/* Enable interrupt before writing */
+	ret = gpio_pin_interrupt_configure_dt(&cfg->status_gpio, GPIO_INT_EDGE_FALLING);
+	if (ret != 0) {
+		LOG_ERR("Failed to configure interrupt: %d", ret);
+		return ret;
+	}
+#endif
+
+	/* Write range command to sensor */
+	ret = i2c_write_dt(&cfg->i2c, &cmd, 1);
+	if (ret != 0) {
+		LOG_ERR("I2C write failed with error %d", ret);
+		return ret;
+	}
+
+	ret = k_sem_take(&data->read_sem, K_MSEC(CONFIG_MB7040_DELAY_MS));
+	if (ret != 0) {
+		if (ret == -EAGAIN) {
+			LOG_ERR("Semaphore take timed out");
+			return -ETIMEDOUT;
+		} else {
+			LOG_ERR("Semaphore take failed with error %d", ret);
+			return ret;
+		}
+	}
+
+#if MB7040_HAS_STATUS_GPIO
+	gpio_pin_interrupt_configure_dt(&cfg->status_gpio, GPIO_INT_DISABLE);
+#endif
+
+	/* Small wait due to devie specific i2c timings */
+	k_msleep(10);
+	ret = i2c_read_dt(&cfg->i2c, read_data, 2);
+	if (ret != 0) {
+		LOG_ERR("I2C read failed with error %d", ret);
+		return ret;
+	}
+
+	/* Convert MSB/LSB to distance in cm */
+	data->distance_cm = (read_data[0] << 8) | read_data[1];
+	return 0;
+}
+
+static int mb7040_channel_get(const struct device *dev, enum sensor_channel chan,
+			      struct sensor_value *val)
+{
+	struct mb7040_data *data = (struct mb7040_data *)dev->data;
+
+	if (chan != SENSOR_CHAN_DISTANCE && SENSOR_CHAN_ALL != chan) {
+		LOG_ERR("Sensor only supports distance");
+		return -EINVAL;
+	}
+
+	val->val1 = data->distance_cm;
+	val->val2 = 0;
+	return 0;
+}
+
+static DEVICE_API(sensor, mb7040_api) = {
+	.sample_fetch = mb7040_sample_fetch,
+	.channel_get = mb7040_channel_get,
+};
+
+static int mb7040_init(const struct device *dev)
+{
+	const struct mb7040_config *cfg = (struct mb7040_config *)dev->config;
+	struct mb7040_data *data = (struct mb7040_data *)dev->data;
+
+	k_sem_init(&data->read_sem, 0, 1);
+
+	if (!i2c_is_ready_dt(&cfg->i2c)) {
+		LOG_ERR("I2C not ready!");
+		return -ENODEV;
+	}
+
+	/* Initialize status GPIO if present */
+#if MB7040_HAS_STATUS_GPIO
```

Comment:
It seems this compile time detection of status GPIO will fail in case of 2 devices with and without such GPIO. I'd recommend using runtime detection of this feature by checking whether `cfg->status_gpio.port` is `NULL` or not.